server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java

```diff
@@ -2246,6 +2249,103 @@ public void testInternalDataStream() {
+            .setDataStreamOptions(DataStreamOptions.FAILURE_STORE_DISABLED)
+            .build();
+        DataStream dataStreamFailureStoreExplicitlyEnabled = createTestInstance().copy()
+            .setName("my-data-stream-failure-store-explicitly-enabled")
+            .setDataStreamOptions(DataStreamOptions.FAILURE_STORE_ENABLED)
+            .build();
+        DataStreamFailureStoreSettings matchingSettings = DataStreamFailureStoreSettings.create(
+            ClusterSettings.createBuiltInClusterSettings(
+                Settings.builder()
+                    .put(DataStreamFailureStoreSettings.DATA_STREAM_FAILURE_STORED_ENABLED_SETTING.getKey(), "my-data-stream-*")
+                    .build()
+            )
+        );
+        DataStreamFailureStoreSettings nonMatchingSettings = DataStreamFailureStoreSettings.create(
+            ClusterSettings.createBuiltInClusterSettings(
+                Settings.builder()
+                    .put(DataStreamFailureStoreSettings.DATA_STREAM_FAILURE_STORED_ENABLED_SETTING.getKey(), "not-my-data-stream-*")
+                    .build()
+            )
+        );
+        assertThat(dataStreamNoFailureStoreOptions.isFailureStoreEffectivelyEnabled(matchingSettings), is(true));
+        assertThat(dataStreamNoFailureStoreOptions.isFailureStoreEffectivelyEnabled(nonMatchingSettings), is(false));
+        assertThat(dataStreamFailureStoreExplicitlyDisabled.isFailureStoreEffectivelyEnabled(matchingSettings), is(false));
+        assertThat(dataStreamFailureStoreExplicitlyDisabled.isFailureStoreEffectivelyEnabled(nonMatchingSettings), is(false));
+        assertThat(dataStreamFailureStoreExplicitlyEnabled.isFailureStoreEffectivelyEnabled(matchingSettings), is(true));
+        assertThat(dataStreamFailureStoreExplicitlyEnabled.isFailureStoreEffectivelyEnabled(nonMatchingSettings), is(true));
+    }
+
+    public void testIsFailureStoreEffectivelyEnabled_staticHelperMethod() {
+        String name = "my-data-stream";
+        String dotPrefixedName = ".my-dot-prefixed-data-stream";
+        DataStreamFailureStoreSettings matchingSettings = DataStreamFailureStoreSettings.create(
+            ClusterSettings.createBuiltInClusterSettings(
+                Settings.builder()
+                    .put(
+                        DataStreamFailureStoreSettings.DATA_STREAM_FAILURE_STORED_ENABLED_SETTING.getKey(),
+                        String.join(",", name, dotPrefixedName)
+                    )
+                    .build()
+            )
+        );
+        DataStreamFailureStoreSettings nonMatchingSettings = DataStreamFailureStoreSettings.create(
+            ClusterSettings.createBuiltInClusterSettings(
+                Settings.builder()
+                    .put(DataStreamFailureStoreSettings.DATA_STREAM_FAILURE_STORED_ENABLED_SETTING.getKey(), "not-my-data-stream")
+                    .build()
+            )
+        );
+        SystemIndices systemIndices = EmptySystemIndices.INSTANCE;
```

Comment:
Yeah, I'd initially tried to do that, and run into the problem that `SystemDataStreamDescriptor` won't like you set a non-dot-prefixed name, making it impossible to test the system data stream logic independently of the dot-prefix logic — the method would return false for `.some-system-data-stream` whether or not the `SystemIndices` object reckoned it was a system data stream or not.

By that reasoning, we could actually dispense with the `SystemIndices` check in the code under test, but it seems like good defensive programming to do it anyway.

But it occurs to me that I can mock `SystemIndices` and get it to behave in the way that the real one never will. See what you think.

Reply:
> By that reasoning, we could actually dispense with the SystemIndices check in the code under test, but it seems like good defensive programming to do it anyway. 

That is probably there because we might have legacy indices, but data streams are pretty new so I don't think we will encounter this. That being said, I prefer to have it for completeness for now.

I like the idea of mocking if it is not too much trouble.